contracts/Synthswap.sol

```diff
@@ -5,85 +5,304 @@ import "./interfaces/ISynthSwap.sol";
 import "./interfaces/IERC20.sol";
 import "./interfaces/ISynthetix.sol";
 import "./interfaces/IAddressResolver.sol";
+import "./interfaces/IAggregationRouterV4.sol";
+import "./interfaces/IAggregationExecutor.sol";
+import "./utils/SafeERC20.sol";
+import "./utils/Owned.sol";
+import "./utils/ReentrancyGuard.sol";
+import "./libraries/RevertReasonParser.sol";
 
-contract SynthSwap is ISynthSwap {
-    ISynthetix synthetix;
-    IERC20 sUSD;
-    IAddressResolver addressResolver;
-    address volumeRewards;
-    address aggregationRouterV3;
+/// @title system to swap synths to/from many erc20 tokens
+/// @dev IAggregationRouterV4 relies on calldata generated off-chain
+contract SynthSwap is ISynthSwap, Owned, ReentrancyGuard {
+    using SafeERC20 for IERC20;
+
+    IERC20 immutable sUSD;
+    IAggregationRouterV4 immutable router;
+    IAddressResolver immutable addressResolver;
+    address immutable volumeRewards;
+    address immutable treasury;
+
+    address constant ETH_ADDRESS = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+    bytes32 private constant CONTRACT_SYNTHETIX = "Synthetix";
+    bytes32 private constant sUSD_CURRENCY_KEY = "sUSD";
+    bytes32 private constant TRACKING_CODE = "KWENTA";
 
     event SwapInto(address indexed from, uint amountReceived);
     event SwapOutOf(address indexed from, uint amountReceived);
+    event Received(address from, uint amountReceived);
     
-    constructor (address _synthetix, address _sUSD, address _volumeRewards, address _aggregationRouterV3, address _addressResolver) {
-        synthetix = ISynthetix(_synthetix);
+    constructor (
+        address _owner,
+        address _sUSD,
+        address _aggregationRouterV4,
+        address _addressResolver,
+        address _volumeRewards,
+        address _treasury
+    ) Owned(_owner) {
```

Comment:
Honestly I would just make the owner `msg.sender` because we can always transfer it from the deployer later. Otherwise this PR is good to go.